src/Service/Storage/Ef/EfTenantStorage.cs

```diff
@@ -285,6 +285,31 @@ public Task RemoveAuthenticatorsAsync(IEnumerable<Guid> aaGuids)
             .ExecuteDeleteAsync();
     }
 
+    public async Task<DispatchedEmail> AddDispatchedEmailAsync(string userId, string emailAddress)
+    {
+        var email = new DispatchedEmail
+        {
+            Tenant = Tenant,
+            Id = Guid.NewGuid(),
+            CreatedAt = _timeProvider.GetUtcNow(),
+            UserId = userId,
+            EmailAddress = emailAddress
+        };
+
+        db.DispatchedEmails.Add(email);
+        await db.SaveChangesAsync();
+
+        return email;
+    }
+
+    public async Task<int> GetDispatchedEmailCountAsync(TimeSpan duration)
+    {
+        var from = _timeProvider.GetUtcNow().UtcDateTime - duration;
+        return await db.DispatchedEmails
+            .Where(x => x.CreatedAt >= from)
```

Comment:
This query could get quite expensive quickly. MSSQL is not so good at creating a good query plan over a time range.

We could consider a view instead that would only be on SQL and it would be better and determining the plan. We could then query one record by application containing it's current minute total and it's current monthly total. Maybe an early optimization but I just get worried about querying against time spans. Especially since I don't see a delete after an amount of time.

Reply:
Added a top-level comment based on our discussion